### ISSUE TITLE
Do shap values from model_output=logloss sum exactly to logloss?

### ISSUE THREAD

**Issue by jphall663** [shap/shap#380]
Using:

`explainer = shap.TreeExplainer(xgb_model, test_yhat[X], feature_dependence='independent', model_output='logloss')` 

I am able to calculate shap contributions to the logloss (or something like that) for each row, but I am unable to work out how these shap values sum to the logloss *exactly*. Maybe they are not supposed to be exact? Maybe I am doing something dumb?

*** 

Here is the setup: 

ubuntu v. 16.04
python v. 3.6.4
shap v. 0.26.0
xgboost v. 0.7

For a given row, the label, prediction, and logloss (calculated by me) is:

y | p_DEFAULT_NEXT_MONTH | r_DEFAULT_NEXT_MONTH
-- | -- | --
1 | 0.028777 | 3.548167

I calculate logloss with: 

`test_yhat['r_DEFAULT_NEXT_MONTH'] = -test_yhat[y]*np.log(test_yhat['p_DEFAULT_NEXT_MONTH']) -\
                                     (1 - test_yhat[y])*np.log(1 - test_yhat['p_DEFAULT_NEXT_MONTH'])`

The Shapley contributions to logloss for the inputs are:

Input | Contrib
--|--
PAY_AMT2   |  0.444923
PAY_3      |  0.372438
PAY_0     |   0.311950
PAY_AMT3  |   0.275476
PAY_2    |    0.217491
PAY_AMT1   |  0.182761
PAY_6     |   0.166236
LIMIT_BAL  |  0.141369
PAY_5    |    0.139306
PAY_AMT5   |  0.135913
BILL_AMT1  |  0.115837
PAY_AMT4   |  0.086789
MARRIAGE  |   0.086296
PAY_4    |    0.064933
BILL_AMT5  |  0.045914
AGE    |      0.038773
BILL_AMT2  |  0.032538
PAY_AMT6   |  0.030993
BILL_AMT4  |  0.024876
BILL_AMT6  |  0.009982
BILL_AMT3  | -0.009435
EDUCATION  |  -0.033935
SEX    |     -0.074084

The sum of these is: `2.8073386454201277`

`explainer.expected_value` is `0.319597019754289`

*** 

I'm not seeing how to combine `2.8073386454201277` and `0.319597019754289` to make `3.548167` ... any assistance would be greatly appreciated!

**Comment by slundberg** [comment]
Hmm...they should sum up exactly. Could you share a notebook I could use to debug it?

**Comment by jphall663** [comment]
Notebook: https://github.com/jphall663/interpretable_machine_learning_with_python/blob/ph_add_errors/shap_loss_test.ipynb

(@slundberg - is there an example for this function? I couldn't find one ... if not, please feel free to use this code to make one, or I can commit once it is working correctly?)

**Comment by slundberg** [comment]
Thanks! I dug around yesterday and found an issue that I am now trying to fix with it. Once I get that done I'll see if it fixes what you ran into.

I have not yet posted an example. It looks like it still needs a bit more testing, but I'll post examples and do a writeup once I am confident in the implementation being right :)

**Comment by slundberg** [comment]
...just an update. It looks like differences can show up right now because of numerical precision issues. If XGBoost decides a value is less than a threshold based on how it quantizes to 32 bit floats then it is very hard to match that, since the dumped version of the model doesn't seem to quantize the same way in Python in some cases. This can lead to a different decision path, and I am still looking into how to fix it. (not sure if this is the issue that came up for you yet)

**Comment by jphall663** [comment]
Thanks for update 👍

**Comment by slundberg** [comment]
I have fixed the precision issues, and also an issue with the expected_value when explaining the loss. My guess is that you were running into the expected_value issue. Since the expected value is taken over the input to the function, it will change if you are comparing the model output to different labels for the loss calculation. So in other words the expected_value depends on the label when you explaining the loss. To see an example check out: https://github.com/slundberg/shap/blob/master/notebooks/tree_explainer/Explaining%20the%20Loss%20of%20a%20Model.ipynb

Note that all these fixes are only on master right now.

**Comment by jphall663** [comment]
Excellent, my test case is passing now too.

The one catch I ran into is: how do I pass the number of trees into the explainer? When I tried the "normal way": `explainer = shap.TreeExplainer(xgb_model, test_yhat[X], feature_dependence='independent', model_output='logloss')` this ran on the model with all trees, not accounting for early stopping, which is slower and also won't correspond to the residuals from the best model. Maybe this is an RTFM question ... sorry if so. I also don't think this counts as a bug, but maybe document it somewhere ... it took me about an hour to figure out why the shap contribs still weren't adding up. (It may also be because I use the direct XGBoost API, not scikit-learn?)

**Comment by slundberg** [comment]
Yeah this is probably more of a I should "write the manual" question :) ...you probably found `tree_limit` already as an option to the shap_values function. But I bet that the expected value does not account for that...so I'll need to update that.
